primitives/transaction/src/historic_transaction.rs

```diff
@@ -0,0 +1,274 @@
+        } else {
+            unreachable!()
+        }
+    }
+
+    /// Returns the hash of the underlying transaction/inherent. For reward inherents we return the
+    /// hash of the corresponding reward transaction. This results into an unique hash for the
+    /// reward inherents (which wouldn't happen otherwise) and allows front-end to fetch rewards by
+    /// their transaction hash.
+    pub fn tx_hash(&self) -> Blake2bHash {
+        match &self.data {
+            HistoricTransactionData::Basic(tx) => tx.hash(),
+            // TODO: check for freedom of hash prefixes
+            _ => self.data.hash(),
+        }
+    }
+
+    /// Tries to convert an historic transaction into a regular transaction. This will work for all
+    /// historic transactions that wrap over regular transactions and reward inherents.
+    pub fn into_transaction(self) -> Result<ExecutedTransaction, IntoTransactionError> {
+        match self.data {
+            HistoricTransactionData::Basic(tx) => Ok(tx),
+            HistoricTransactionData::Reward(ev) => {
+                Ok(ExecutedTransaction::Ok(BlockchainTransaction::new_basic(
+                    Policy::COINBASE_ADDRESS,
+                    ev.reward_address,
+                    ev.value,
+                    Coin::ZERO,
+                    self.block_number,
+                    self.network_id,
+                )))
+            }
+            HistoricTransactionData::Penalize(_)
+            | HistoricTransactionData::Jail(_)
+            | HistoricTransactionData::Equivocation(_) => {
+                Err(IntoTransactionError::NoBasicTransactionMapping)
+            }
+        }
+    }
+}
+
+impl MMRHash<Blake2bHash> for HistoricTransaction {
+    /// Hashes a prefix and an historic transaction into a Blake2bHash. The prefix is necessary
+    /// to include it into the History Tree.
+    fn hash(&self, prefix: u64) -> Blake2bHash {
+        let mut message = prefix.to_be_bytes().to_vec();
+        message.append(&mut self.serialize_to_vec());
+        message.hash()
+    }
+}
+
+impl IntoDatabaseValue for HistoricTransaction {
+    fn database_byte_size(&self) -> usize {
+        self.serialized_size()
+    }
+
+    fn copy_into_database(&self, mut bytes: &mut [u8]) {
+        Serialize::serialize_to_writer(&self, &mut bytes).unwrap();
+    }
+}
+
+impl FromDatabaseValue for HistoricTransaction {
+    fn copy_from_database(bytes: &[u8]) -> io::Result<Self>
+    where
+        Self: Sized,
+    {
+        Deserialize::deserialize_from_vec(bytes)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
+    }
+}
+
+/// An enum specifying the type of transaction and containing the necessary data to represent that
+/// transaction.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, SerializeContent)]
+pub enum HistoricTransactionData {
+    /// A basic transaction. It simply contains the transaction as contained in the block.
+    Basic(ExecutedTransaction),
```

Comment:
I think calling all transactions (as opposed to inherents) `Basic` is misleading, as transactions themselves are also split into 'basic' and 'extended' formats, and this `Basic` here represents both of those formats.

Reply:
Maybe "normal transaction" would work?

Note that this pull request didn't introduce this terminology, it just kept using it.

Reply:
Yes, I already noticed it a while ago.

Maybe just 'transaction' vs 'inherent'?